### PR TITLE
Add search bar for top level settings in Manage Jenkins

### PR DESCRIPTION
**Pull request by janfaracik**
https://user-images.githubusercontent.com/43062514/198851674-054336f6-0f6f-4473-990f-131f53669513.mov

This PR adds a small search bar which searches the top level (just visible) settings on the Manage Jenkins page, it's pretty simple in it's featureset but it makes Jenkins a little easier to navigate via keyboard.

It also improves the search bar component, it's now possible to provide it a list of items and the search bar will provide suggestions from that set of data - this will be added to the Design Library.

It would be great to have this extended in future to include items from each settings page, but for now I felt that making it more keyboard accessible + improving the search bar component was enough.

---

image

One of the things noticed whilst building this is the unnecessary prefixes, e.g. '_Configure System', '_Manage_ Plugins', '_Manage_ Nodes and Clouds' - you're already in Manage Jenkins so it seems unnecessary to include the word 'Configure/Manage' in many of the items. I'd propose dropping the word from each item in the future + this would make search results easier to read, any thoughts?

### Testing done

* Open up Manage Jenkins, search should autofocus, type to filter items on the page. The only item that won't appear is 'Reload configuration from disk' as this is an action, not a link.
* Use arrow keys (up/down) to focus different items, press enter key to navigate to the currently selected one.
* Open up a page with the search bar e.g. a build, plugin manager (not Manage Jenkins) and check existing functionality still works, check console to see that there are no errors

### Proposed changelog entries

- Add search bar for top level settings in Manage Jenkins.

### Proposed upgrade guidelines

N/A

### Developer changelog

- The search bar component now displays search suggestions when passed a dataset (see Jenkins Design Library)

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7314"></a>

### CODE REVIEW

**Comment by NotMyFault** [comment]
I'm really fond of this feature!

Menu entry highlighting doesn't look too good with dark mode though, if you can notice the barely visible frame on the second action item from the top.

Screenshot 2022-10-29 at 23 34 38

**Comment by DuMaM** [comment]
Great improvement. 
I wish main search bar will allow for such things ✨ 👀

**Review by NotMyFault — APPROVED** [review]

**Comment by Kevin-CB** [comment]
I did a review, from a security standpoint, I've no problem with this request.

But I was able to notice behaviors that I don't think were intended:

- When you start a search, click somewhere else on the page, then come back to your search, it does not expand completely.

image

----

- If you've warnings with buttons, they will not be totally hidden by the search box, leaving blurred blue spots.

image

**Comment by janfaracik** [comment]
> * When you start a search, click somewhere else on the page, then come back to your search, it does not expand completely.
> 
> image

Hmm, I can't reproduce that - what browser are you using?

> * If you've warnings with buttons, they will not be totally hidden by the search box, leaving blurred blue spots.
> 
> image

The blurring behind the results is intentional (like frosted glass) but the blue is definitely a bit too vibrant behind it 👍

**Comment by Kevin-CB** [comment]
> Hmm, I can't reproduce that - what browser are you using?

Firefox, I also tried with Chrome and Edge and I'm not able to reproduced it with them.

File: core/src/main/resources/lib/layout/search-bar.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/search-bar.jelly:41]: Why is this a `jenkins-` and not an `app`? Do we expect plugins to copy & paste that? Is this stable API? I assume not, given we're changing a previously `jenkins-` prefixed element structure to support an unrelated UI element.

**Review by daniel-beck — COMMENTED** [review]
First impressions: I like the idea a lot.

When using it, I was at first surprised that I didn't find results when specifying text content from the description rather than the title. Depending on the search terms, this might quickly get messy though.

I do not think it is a good idea to add yet another design of a dropdown to Jenkins (very rounded corners, light background, with icons), in addition to configuration form hetero dropdowns (with filter, no icons), the global search dropdown (slightly rounded corners, bold highlight text, dark blue background, no icons), and context menus (corners, medium blue background, with icons). If you don't like what dropdown/popup menus look like, change the design for all of them in a separate change.

**Comment by janfaracik** [comment]
> When using it, I was at first surprised that I didn't find results when specifying text content from the description rather than the title. Depending on the search terms, this might quickly get messy though.

Yeah I was worried it might get quite muddled if we have provide too much data to query on, definitely would be a worry too if we were to start querying settings pages too.

> I do not think it is a good idea to add yet another design of a dropdown to Jenkins

> global search dropdown (slightly rounded corners, bold highlight text, dark blue background, no icons)

I've held off completely on updating the navigation search bar as I'm hoping to replace it with a more command palette-esque interface which will feature a UI inline with this one (albeit larger).

image

> configuration form hetero dropdowns (with filter, no icons)
> context menus (corners, medium blue background, with icons)

I'll be starting on updating the menus/dropdowns once Tippy is merged in, so the dropdowns on Jenkins should be consistent once that's done.

---

Thanks! :)

File: core/src/main/resources/lib/layout/search-bar.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/search-bar.jelly:41]: I would expect users to use the search-bar component rather than cobbling it together themselves. It is a fair question though, in this case though any existing implementations with the class ".jenkins-search" would still work as it's just the container that's been added so that the search results are positioned correctly.

**Comment by timja** [comment]
Seems to be a bit odd to autofocus something when it's off in the top right.

image

should either be central (or maybe left sidepanel) or not autofocus I think just seems weird to draw my attention over there

File: war/src/main/js/components/search-bar/index.js
Comment (Kevin-CB) [inline war/src/main/js/components/search-bar/index.js:39]: Could you escape the `item.label` variable? It's not dangerous here, since the values are hardcoded, but this code could be reused for other pages with search capabilities, where the labels might be dangerous and thus introduce an XSS.

Comment (janfaracik) [reply war/src/main/js/components/search-bar/index.js]: Good idea! Done.

File: war/src/main/js/util/security.js
Comment (Kevin-CB) [inline war/src/main/js/util/security.js:2]: It won't work, if we take the example of the Russian, Chinese, Japanese alphabet, etc., nothing will be displayed.

The best approach will be to convert dangerous characters into their HTML encoded version, you can find an example with the function [Util.xmlEscape](https://github.com/jenkinsci/jenkins/blob/d3d1c88d299735186209ad5b2f7a41e05486105f/core/src/main/java/hudson/Util.java#L1045).

**Review by Kevin-CB — APPROVED** [review]
LGTM 🙂

**Review by timja — APPROVED** [review]
LGTM re-tested it's nice, I agree we should remove the repeated configure where it makes sense but in a different PR 😄 

---

/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
